book-content/chapters/11-annotations-and-assertions.md

```diff
@@ -0,0 +1,1196 @@
+
+```tsx
+const album = {
+  format: "Vinyl",
+};
+```
+
+Here, we have an `album` object with a `format` key. As we know from our chapter on mutability, TypeScript will infer `album.format` as `string`. We want to make sure that the `format` is one of three values: `CD`, `Vinyl`, or `Digital`.
+
+We could give it a variable annotation:
+
+```tsx
+type Album = {
+  format: "CD" | "Vinyl" | "Digital";
+};
+
+const album: Album = {
+  format: "Vinyl",
+};
+```
+
+But now, `album.format` is `"CD" | "Vinyl" | "Digital"`. This might be a problem if we want to pass it to a function that only accepts `"Vinyl"`.
+
+Instead, we can use `satisfies`:
+
+```typescript
+const album = {
+  format: "Vinyl",
+} satisfies Album;
+```
+
+Now, `album.format` is inferred as `"Vinyl"`, because we've told TypeScript that `album` satisfies the `Album` type. So, `satisfies` is narrowing down the value of `album.format` to a specific type.
+
+## Assertions: Forcing The Type Of Values
+
+Sometimes, the way TypeScript infers types isn't quite what we want. We can use assertions in TypeScript to force values to be inferred as a certain type.
+
+### The `as` Assertion
+
+The `as` assertion is a way to tell TypeScript that you know more about a value than it does. It's a way to override TypeScript's type inference and tell it to treat a value as a different type.
+
+Let's look at an example.
+
+Imagine that you're building a web page that has some information in the search query string of the URL.
+
+You happen to know that the user can't navigate to this page without passing `?id=some-id` to the URL.
+
+```ts twoslash
+const searchParams = new URLSearchParams(window.location.search);
+
+const id = searchParams.get("id");
+//    ^?
+```
+
+But TypeScript doesn't know that the `id` will always be a string. It thinks that `id` could be a string or `null`.
+
+So, let's force it. We can use `as` on the result of `searchParams.get("id")` to tell TypeScript that we know it will always be a string:
+
+```ts twoslash
+const searchParams = new URLSearchParams(window.location.search);
+// ---cut---
+const id = searchParams.get("id") as string;
+//    ^?
+```
+
+Now TypeScript knows that `id` will always be a string, and we can use it as such.
+
+This `as` is a little unsafe! If `id` is somehow not actually passed in the URL, it will be `null` at runtime but `string` at compile time. This means if we called `.toUpperCase()` on `id`, we'd crash our app.
+
+But it's useful in cases where we truly know more than TypeScript can about the behavior of our code.
+
+#### An Alternative Syntax
+
+As an alternative to `as`, you can prefix the value with the type wrapped in angle brackets:
+
+```typescript
+const id = <string>searchParams.get("id");
+```
+
+This is less common than `as`, but behaves exactly the same way. `as` is more common, so it's better to use that.
+
+#### The Limits of `as`
+
+`as` has some limits on how it can be used. It can't be used to convert between unrelated types.
```

Comment:
It can only be used to "convert" (downcast or upcast) between **comparable** types (one has to be a subtype of the other one for it to work)... with a "beautiful" exception of:
```ts
declare const b: "b"
const a = "a" as typeof b;
```

Those shouldn't quite be comparable - two literal types don't form the subtype/supertype relationship and yet this casting is allowed. I think I read that this was a bug that slipped through for enough period of time that it has now been codified as a feature :v